src/engine/Map.cxx

```diff
@@ -428,34 +426,42 @@ Point Map::findNodeInMap(const SDL_Point &screenCoordinates) const
     isoY -= diff;
   }
 
-  // traverse a column from top to bottom (from the calculated coordinates) our calculated point is always higher than the clicked point
-  while (isoX <= Settings::instance().mapSize && isoY <= Settings::instance().mapSize && isoY >= 0)
+#ifdef DEBUG
+  int zOrder = INT_MAX;
+#endif
+  // Transerse a column form from calculated coordinates to the bottom of the map.
+  // It is necessary to include 2 neighbor nodes from both sides.
+  // Try to find map node in Z order.
+  // Node with the highest Z order has highest X and lowest Y coordinate, so search will be conducted in that order.
+  const int neighborReach = 2;
+  const int mapSize = Settings::instance().mapSize;
+
+  // Max X will reach end of the map or Y will reach 0.
+  const int xMax = std::min(isoX + neighborReach + isoY, mapSize - 1);
+  // Min X will reach 0 or x -2 neighbor node.
+  const int xMin = std::max(isoX - neighborReach, 0);
+
+  for (int x = xMax; x >= xMin; --x)
   {
-    // include 2 columns on each side, since calculated values can be that far off
-    for (int i = 0; i <= 2; i++)
+    const int diff = x - isoX;
+    const int yMiddlePoint = isoY - diff;
+
+    // Move y up and down 2 neighbors.
+    for (int y = std::max(yMiddlePoint - neighborReach, 0); (y <= yMiddlePoint + neighborReach) && (y < mapSize); ++y)
     {
-      if (isClickWithinTile(screenCoordinates, isoX, isoY) &&
-          (foundCoordinates.z < mapNodes[isoX * m_columns + isoY]->getCoordinates().z))
-      {
-        foundCoordinates = mapNodes[isoX * m_columns + isoY]->getCoordinates();
-      }
-      if (isClickWithinTile(screenCoordinates, isoX + i, isoY) &&
-          (foundCoordinates.z < mapNodes[(isoX + i) * m_columns + isoY]->getCoordinates().z))
-      {
-        foundCoordinates = mapNodes[(isoX + i) * m_columns + isoY]->getCoordinates();
-      }
-      if (isClickWithinTile(screenCoordinates, isoX, isoY - i) &&
-          (foundCoordinates.z < mapNodes[isoX * m_columns + (isoY - i)]->getCoordinates().z))
+#ifdef DEBUG
+      // Assert asumption that we test all nodes in correct Z order
+      assert(zOrder > mapNodes[x * m_columns + y]->getCoordinates().z);
```

Comment:
Why not write a proper unit test for it instead of having asserts in debug mode?

Reply:
This will test all cases during run-time, which is very unlikely to achieve with unit tests. Also will notify if someone changes Z-order handing in terain generator, on which assumption this algorithm is based. In release code version this will impose no overhead at all.